test/dynamo/test_aot_autograd.py

```diff
@@ -1412,6 +1412,47 @@ def forward(self, x):
         ):
             out.backward(retain_graph=True)
 
+    def test_inputs_overlapping_with_mutation_recompile(self):
+        def f(*args):
+            for a in args:
+                a.add_(1)
+            return args[0]
+
+        def overlapping_args(x):
+            return x[:5], x[7:13], x[9:]
+
+        def non_overlapping_args(x):
+            return x[:5], x[7:13], x[13:15]
+
+        guard_failure = []
+
+        def guard_fail_fn(failure):
+            nonlocal guard_failure
+            guard_failure.append(failure[0])
+
+        input = torch.ones(20)
+        opt_input = input.clone().detach()
+
+        opt_f = torch._dynamo.optimize(
+            "aot_eager", dynamic=True, guard_fail_fn=guard_fail_fn
+        )(f)
+
+        out0 = f(*overlapping_args(input))
+        opt_out0 = opt_f(*overlapping_args(opt_input))
+        self.assertEqual(out0, opt_out0)
+
+        out1 = f(*non_overlapping_args(input))
+        opt_out1 = opt_f(*non_overlapping_args(opt_input))
+        self.assertEqual(out1, opt_out1)
```

Comment:
A few more things that would be nice to test for:

(1) a stress test where you have a function with a large number of inputs (100+), that asserts that the total number of guards / shape guards is not large (to test that we never accidentally end up generating quadratic shape guards at some point

(2) in terms of recompiles: right now this test properly tests that we recompile when we go from "some overlapping inputs" to "no overlapping inputs". Can we also test that we properly recompile whenever the exact set of overlapping indices changes? (mostly to confirm that the bug mentioned from the last round of feedback would have been caught)

(3) this guard is expensive, so we want to make sure we generate it only when we need to. Can you add a test that we e.g. do not generate the guard, when you have no mutations in the graph, or when a mutated input has no other aliases?